bad/source/lib.rs

```diff
@@ -11,31 +17,47 @@ pub enum ProgramSource {
 }
 
 pub struct CompilationConfiguration {
-	pub sources: Vec<ProgramSource>,
+	
 }
 
 #[derive(Debug)]
-pub enum Token {
-	LeftParen,
-	RightParen,
-	LeftBrace,
-	RightBrace,
-	Semicolon,
-	SingleQuote,
-	Codepoint(char),
-	Identifier(String),
+pub struct TokenList {
+	pub tokens: Vec<lex::Token>,
+	pub context: ast::Context,
 }
 
-#[derive(Debug)]
-pub struct TokenList {
-	pub tokens: Vec<Token>,
+impl TokenList {
+	pub fn new(path: PathBuf, source: String) -> TokenList {
+		Self {
+			tokens: Vec::new(),
+			context: ast::Context::new(path, source)
+		}
+	}
 }
 
 #[derive(Debug)]
 pub struct SyntaxTree {}
 
-pub fn lex(_config: &CompilationConfiguration) -> TokenList {
-	let list: TokenList = TokenList { tokens: Vec::new() };
+pub fn get_source_text(source: ProgramSource) -> (PathBuf, String) {
+	let mut source_text: String = String::new();
+	let mut source_path: PathBuf = PathBuf::new();
+	match source {
+		ProgramSource::Path(pathbuf) => { source_path = pathbuf.clone(); source_text = std::fs::read_to_string(pathbuf.as_path()).unwrap() },
+		ProgramSource::File(mut file) => { source_path = "<file?handle?>".into(); file.read_to_string(&mut source_text); },
+		ProgramSource::Stdin(mut stdin) => { source_path = "<stdin>".into(); stdin.read_to_string(&mut source_text); }
+	};
+	(source_path, source_text)
+}
+
+pub fn lex(source_path: PathBuf, source_text: String, config: &CompilationConfiguration) -> TokenList {
+	let mut list: TokenList = TokenList::new(source_path, source_text);
+	let logos_lex = Token::lexer(list.context.source()).spanned();
+	for tok_and_range in logos_lex {
```

Comment:
It might be more idiomatic to destrucutre in the `for` loop, like this:

```rs
for (tok, range) in lex {
  match tok { .,. }
}
```

Avoids repeating _range everywhere.

Reply:
Okay!